src/collection/backend/lmdb.cc

```diff
@@ -503,6 +501,62 @@ void LMDB::resolveRegularExpression(const std::string& var,
     return;
 }
 
+
+MDBEnvProvider* MDBEnvProvider::provider_ = nullptr;
+
+MDBEnvProvider* MDBEnvProvider::GetInstance() {
+    if (provider_==nullptr) {
+        provider_ = new MDBEnvProvider();
+    }
+    return provider_;
+}
+
+void MDBEnvProvider::Finalize() {
+    if (provider_!=nullptr) {
+        provider_->close();
+        provider_ = nullptr;
+    }
+}
+
+MDBEnvProvider::MDBEnvProvider() :
+    m_env(NULL), initialized(false) {
+    pthread_mutex_init(&m_lock, NULL);
```

Comment:
Just wondering if this might be better inside the if-block at line 538 in init().

As it is, it looks like it at least be theoretically possible to have a thread-safety issue here where pthread_mutex_init could get called on an already-initialized m_lock.

Reply:
Right. My earlier comments weren't fully thought through.

However, I do still wonder about a sequence of events like this:
- thread 1:
 - calls GetInstance and runs to line 508
 - line 508 is executed, the boolean expression is evaluated to true
 - execution advances to the beginning of 509.
thread 2:
 - calls GetInstance and runs to line 508
 - line 508 is executed, the boolean expression is again evaluated to true
 - execution advances all the way to the end of GetInstance(); an MDBEnvProvider has been constructed -- let's call it MDBEnvProvider object A
 - MDBEnvProvider object A is returned to the caller of GetInstance
thread 1:
 - completes executing GetInstance; a different new MDBEnvProvider is constructed ('MDBEnvProvider object B')
 - MDBEnvProvider object B is returned to the caller of GetInstance

And we have two objects instead of 1. (And now both of these separate MDBEnvProvider objects can have init() called. That would include having mdb_dbi_open called both times -- and this possibility is part of what prompted the original #2601 fix.)

It's been a while since I've done anything with Singletons, so I've had to do a bit of review.

Your mechanism (static member pointer variable, check for NULL, then create the object via new) is the Gang-of-Four mechanism, but it seems not fully thread-safe.

Since the code base is firmly in C++11, might a Meyer's Singleton be a better implementation?

(Apologies if I've mucked something up, as I said, it's been a while since I've used a this Design Pattern.)

Reply:
Thanks for your detailed clarification, I agree entirely. I've made a silent assumption that *GetInstance()* is called only by the main thread. I've been biased by focusing on nginx-connector use case - where this assumption is fulfilled indeed. 

I'll replace this implementation with one that is thread-safe, as You suggested.